docs/User/Tutorials/haskell-lazy-evaluation.md

```diff
@@ -0,0 +1,171 @@
+the length of the list.  Till now Haskell has been saying that I duly
+noted how to do what you want to do but I will do it only when you need
+it. Let's ask it to print `len` now:
+
+```
+>>> len
+...hangs forever
+```
+
+Hard pressed, Haskell now has no choice but to start evaluating the
+expression but we know it can never finish because the list is infinite.
+
+All expressions in Haskell are unevaluated by default.  Haskell
+evaluates the expressions only when it is absolutely required e.g. when
+we have to print something, or write it to a file or to network.
+
+## On-demand Evaluation
+
+Let us illustrate expression evaluation in Haskell with a few examples. Write
+this little program in a file `example.hs`:
+
+```
+main =
+    let largeList = [1..1000000::Int]
+        n = length largeList
+     in print n
+```
+
+Compile and run it:
+
+```
+$ ghc -O example.hs
+[1 of 1] Compiling Main             ( example.hs, example.o )
+Linking example ...
+$ ./example +RTS -s
+1000000
+          51,480 bytes allocated in the heap
+              24 bytes copied during GC
+          44,328 bytes maximum residency (1 sample(s))
+```
+
+`largeList` is an expression representing a million element list, but
+the program takes only 44K maximum memory. The reason is we never evaluate
+and store the entire list in memory.  The length function evaluates
+`largeList` one element at a time to count them, as soon as an element is
+counted it can be garbage collected as it is not used anywhere else in the
+program.
+
+## Retained Heap Memory
+
+Let's try this program instead:
+
+```
+main = do
+    let largeList = [1..1000000::Int]
+        x1 = last largeList
+        x2 = head largeList
+     in print (x1, x2)
+```
+
+```
+$ ./example +RTS -s
+(1000000,1)
+      64,051,552 bytes allocated in the heap
+             400 bytes copied during GC
+      26,233,360 bytes maximum residency (4 sample(s))
+```
+
+This program is showing 26MB maximum residency. Why is it so? To
+evaluate the first element of the tuple, `x1`, we need to evaluate
+the entire list to get to the last element. However, the head of the
+list is referenced by the element `x2` as well. Therefore, we need to
+retain the head until `x2` is printed and freed. But head of the list is
+referencing the entire list, therefore, the entire list is retained in
+memory between the evaluation of `x1` and `x2`.
+
+If we change the print statement to `(x2, x1)` then we do not need to
+retain the entire list between the two.
+
+## Forcing Evaluation
+
+Even if there is no explicit dependency in the program, the programmer can
+force the evaluation of an expression using `seq`:
+
+```
+main = do
+    let largeList = [1..1000000::Int]
+        x1 = last largeList
+        x2 = head largeList
+     in print (x1 `seq` x2)
+```
+
+Similar to the example in previous section, this also retains the entire list
+in memory because `x1` is forced to be evaluated before `x2`.
+
+## Reusing without retaining
```

Comment:
The contents of this seem a little out of place for some reason.

Reply:
This tightly related to the previous section, just showing how the problem discussed there can be avoided.